Bug: leading semicolon after single-line block = wrong indentation

### Environment

Node version: v18.0.0
npm version: v8.6.0
Local ESLint version: Not found
Global ESLint version: v8.16.0 (Currently used)
Operating System: linux 5.10.0-9-amd64

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
	"root": true,
	"env": {"es6": true},
	"rules": {
		"indent": ["error", "tab"],
		"semi": ["warn", "never"],
		"semi-style": ["warn", "first"]
	}
}
```
</details>

```js
if (2 > 1)
	console.log('a')
;[1, 2, 3].forEach(x=>console.log(x))
```


### What did you expect to happen?

(no errors)

### What actually happened?

```
  3:1  error  Expected indentation of 1 tab but found 0  indent
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -916,18 +916,6 @@ module.exports = {
                 }
 
                 offsets.setDesiredOffsets([firstBodyToken.range[0], lastBodyToken.range[1]], lastParentToken, 1);
-
-                /*
-                 * For blockless nodes with semicolon-first style, don't indent the semicolon.
-                 * e.g.
-                 * if (foo) bar()
-                 * ; [1, 2, 3].map(foo)
-                 */
-                const lastToken = sourceCode.getLastToken(node);
-
-                if (node.type !== "EmptyStatement" && astUtils.isSemicolonToken(lastToken)) {
-                    offsets.setDesiredOffset(lastToken, lastParentToken, 0);
-                }
             }
         }
 
@@ -1268,6 +1256,50 @@ module.exports = {
                 addBlocklessNodeIndent(node.consequent);
                 if (node.alternate && node.alternate.type !== "IfStatement") {
                     addBlocklessNodeIndent(node.alternate);
+                }
+            },
+
+            /*
+             * For blockless nodes with semicolon-first style, don't indent the semicolon.
+             * e.g.
+             * if (foo)
+             *     bar()
+             * ; [1, 2, 3].map(foo)
+             *
+             * Traversal into the node sets indentation of the semicolon, so we need to override it on exit.
+             */
+            ":matches(DoWhileStatement, ForStatement, ForInStatement, ForOfStatement, IfStatement, WhileStatement):exit"(node) {
+                let nodesToCheck;
+
+                if (node.type === "IfStatement") {
+                    nodesToCheck = [node.consequent];
+                    if (node.alternate) {
+                        nodesToCheck.push(node.alternate);
+                    }
+                } else {
+                    nodesToCheck = [node.body];
+                }
+
+                for (const nodeToCheck of nodesToCheck) {
+                    const lastToken = sourceCode.getLastToken(nodeToCheck);
+
+                    if (astUtils.isSemicolonToken(lastToken)) {
+                        const tokenBeforeLast = sourceCode.getTokenBefore(lastToken);
+                        const tokenAfterLast = sourceCode.getTokenAfter(lastToken);
+
+                        // override indentation of `;` only if its line looks like a semicolon-first style line
+                        if (
+                            !astUtils.isTokenOnSameLine(tokenBeforeLast, lastToken) &&
+                            tokenAfterLast &&
+                            astUtils.isTokenOnSameLine(lastToken, tokenAfterLast)
+                        ) {
+                            offsets.setDesiredOffset(
+                                lastToken,
+                                sourceCode.getFirstToken(node),
+                                0
+                            );
+                        }
+                    }
                 }
             },
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -6171,6 +6171,16 @@ ruleTester.run("indent", rule, {
         {
             code: unIndent`
                 if (foo)
+                    bar()
+                else if (baz)
+                    qux()
+                ;quux()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
                     if (bar)
                         baz()
                     else
@@ -13020,6 +13030,24 @@ ruleTester.run("indent", rule, {
             `,
             options: [4],
             errors: expectedErrors([4, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                else if (baz)
+                    qux()
+                    ;quux()
+            `,
+            output: unIndent`
+                if (foo)
+                    bar()
+                else if (baz)
+                    qux()
+                ;quux()
+            `,
+            options: [4],
+            errors: expectedErrors([5, 0, 4, "Punctuator"])
         },
         {
             code: unIndent`

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -6082,6 +6082,252 @@ ruleTester.run("indent", rule, {
             `,
             options: [4, { FunctionExpression: { body: 2 }, StaticBlock: { body: 2 } }],
             parserOptions: { ecmaVersion: 2022 }
+        },
+
+        // https://github.com/eslint/eslint/issues/15930
+        {
+            code: unIndent`
+                if (2 > 1)
+                \tconsole.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: ["tab"]
+        },
+        {
+            code: unIndent`
+                if (2 > 1)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo) bar();
+                baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo) bar()
+                ;baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar();
+                baz();
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                ; baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                ;baz()
+                qux()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                ;else
+                    baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                else
+                    baz()
+                ;qux()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                ;qux()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                    else
+                        qux()
+                ;quux()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                    ;
+                baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    ;
+                baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                ;baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo);
+                else
+                    baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    ;
+                else
+                    baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                if (foo)
+                ;else
+                    baz()
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do foo();
+                while (bar)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do foo()
+                ;while (bar)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do
+                    foo();
+                while (bar)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do
+                    foo()
+                ;while (bar)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do;
+                while (foo)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do
+                    ;
+                while (foo)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                do
+                ;while (foo)
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                while (2 > 1)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                for (;;)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                for (a in b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                label: for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
+        },
+        {
+            code: unIndent`
+                label:
+                for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4]
         }
     ],
 
@@ -12623,6 +12869,439 @@ ruleTester.run("indent", rule, {
                 [6, 12, 0, "Identifier"],
                 [7, 4, 0, "Punctuator"]
             ])
+        },
+
+        // https://github.com/eslint/eslint/issues/15930
+        {
+            code: unIndent`
+                if (2 > 1)
+                \tconsole.log('a')
+                \t;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                if (2 > 1)
+                \tconsole.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: ["tab"],
+            errors: expectedErrors("tab", [3, 0, 1, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (2 > 1)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                if (2 > 1)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo) bar();
+                    baz()
+            `,
+            output: unIndent`
+                if (foo) bar();
+                baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Identifier"])
+        },
+        {
+            code: unIndent`
+                if (foo) bar()
+                    ;baz()
+            `,
+            output: unIndent`
+                if (foo) bar()
+                ;baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar();
+                    baz();
+            `,
+            output: unIndent`
+                if (foo)
+                    bar();
+                baz();
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Identifier"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                    ; baz()
+            `,
+            output: unIndent`
+                if (foo)
+                    bar()
+                ; baz()
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                    ;baz()
+                    qux()
+            `,
+            output: unIndent`
+            if (foo)
+                bar()
+            ;baz()
+            qux()
+        `,
+            options: [4],
+            errors: expectedErrors([
+                [3, 0, 4, "Punctuator"],
+                [4, 0, 4, "Identifier"]
+            ])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                    ;else
+                    baz()
+            `,
+            output: unIndent`
+                if (foo)
+                    bar()
+                ;else
+                    baz()
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                else
+                    baz()
+                    ;qux()
+            `,
+            output: unIndent`
+                if (foo)
+                    bar()
+                else
+                    baz()
+                ;qux()
+            `,
+            options: [4],
+            errors: expectedErrors([5, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                    ;qux()
+            `,
+            output: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                ;qux()
+            `,
+            options: [4],
+            errors: expectedErrors([4, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                    else
+                        qux()
+                    ;quux()
+            `,
+            output: unIndent`
+                if (foo)
+                    if (bar)
+                        baz()
+                    else
+                        qux()
+                ;quux()
+            `,
+            options: [4],
+            errors: expectedErrors([6, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    bar()
+                ;
+                baz()
+            `,
+            output: unIndent`
+                if (foo)
+                    bar()
+                    ;
+                baz()
+            `,
+            options: [4],
+            errors: expectedErrors([3, 4, 0, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                ;
+                baz()
+            `,
+            output: unIndent`
+                if (foo)
+                    ;
+                baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 4, 0, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    ;baz()
+            `,
+            output: unIndent`
+                if (foo)
+                ;baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo);
+                    else
+                    baz()
+            `,
+            output: unIndent`
+                if (foo);
+                else
+                    baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Keyword"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                ;
+                else
+                    baz()
+            `,
+            output: unIndent`
+                if (foo)
+                    ;
+                else
+                    baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 4, 0, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                if (foo)
+                    ;else
+                    baz()
+            `,
+            output: unIndent`
+                if (foo)
+                ;else
+                    baz()
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                do foo();
+                    while (bar)
+            `,
+            output: unIndent`
+                do foo();
+                while (bar)
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Keyword"])
+        },
+        {
+            code: unIndent`
+                do foo()
+                    ;while (bar)
+            `,
+            output: unIndent`
+                do foo()
+                ;while (bar)
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                do
+                    foo();
+                    while (bar)
+            `,
+            output: unIndent`
+                do
+                    foo();
+                while (bar)
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Keyword"])
+        },
+        {
+            code: unIndent`
+                do
+                    foo()
+                    ;while (bar)
+            `,
+            output: unIndent`
+            do
+                foo()
+            ;while (bar)
+        `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                do;
+                    while (foo)
+            `,
+            output: unIndent`
+                do;
+                while (foo)
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Keyword"])
+        },
+        {
+            code: unIndent`
+                do
+                ;
+                while (foo)
+            `,
+            output: unIndent`
+                do
+                    ;
+                while (foo)
+            `,
+            options: [4],
+            errors: expectedErrors([2, 4, 0, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                do
+                    ;while (foo)
+            `,
+            output: unIndent`
+                do
+                ;while (foo)
+            `,
+            options: [4],
+            errors: expectedErrors([2, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                while (2 > 1)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                while (2 > 1)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                for (;;)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                for (;;)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                for (a in b)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                for (a in b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                for (a of b)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                label: for (a of b)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                label: for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([3, 0, 4, "Punctuator"])
+        },
+        {
+            code: unIndent`
+                label:
+                for (a of b)
+                    console.log('a')
+                    ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            output: unIndent`
+                label:
+                for (a of b)
+                    console.log('a')
+                ;[1, 2, 3].forEach(x=>console.log(x))
+            `,
+            options: [4],
+            errors: expectedErrors([4, 0, 4, "Punctuator"])
         }
     ]
 });

